CoreAlgorithmPlugins/src/au/gov/asd/tac/constellation/plugins/algorithms/sna/similarity/CosineSimilarityPlugin.java

```diff
@@ -109,94 +109,128 @@ public void edit(final GraphWriteMethods graph, final PluginInteraction interact
+                        final float cosineSimilarity = neighboursMagnitude == 0 ? 0 : neighbourDotProduct / neighboursMagnitude;
+                        SimilarityUtilities.addScoreToGraph(leftVertexWithNeighbours.vertexId, rightVertexWithNeighbours.vertexId, cosineSimilarity);
                     }
-
-                    final BitSet intersection = (BitSet) neighbours[vertexOnePosition].clone();
-                    intersection.and(neighbours[vertexTwoPosition]);
-
-                    if (intersection.cardinality() < minCommonFeatures) {
-                        continue;
-                    }
-
-                    final int vertexOneId = graph.getVertex(vertexOnePosition);
-                    final int[] neighboursOne = neighbourWeights[vertexOnePosition];
-
-                    final int vertexTwoId = graph.getVertex(vertexTwoPosition);
-                    final int[] neighboursTwo = neighbourWeights[vertexTwoPosition];
-
-                    final float neighbourDotProduct = dot(neighboursOne, neighboursTwo);
-                    final float neighboursMagnitude = magnitude(neighboursOne) * magnitude(neighboursTwo);
-                    final float cosineSimilarity = neighboursMagnitude == 0 ? 0 : neighbourDotProduct / neighboursMagnitude;
-                    cosineSimilarities.put(Tuple.create(vertexOneId, vertexTwoId), cosineSimilarity);
                 }
             }
         }
-
-        // update the graph with cosine similarity values
-        SimilarityUtilities.addScoresToGraph(graph, cosineSimilarities, COSINE_SIMILARITY_ATTRIBUTE);
-
         // complete with schema
         PluginExecution.withPlugin(VisualSchemaPluginRegistry.COMPLETE_SCHEMA).executeNow(graph);
     }
-
-    private float dot(final int[] vectorOne, final int[] vectorTwo) {
-        assert vectorOne.length == vectorTwo.length;
-
-        float dot = 0;
-        for (int index = 0; index < vectorOne.length; index++) {
-            dot += vectorOne[index] * vectorTwo[index];
+    
+    private class VertexWithNeighbours {
+        private final int vertexId;
+        private final HashMap<Integer, Integer> neighbourWeightsMap;
+        private BitSet neighbours;
+        private final boolean selected;
+        private Float magnitude = null;
+
+        public VertexWithNeighbours(final int vertexId, final int vertexCount, final boolean selected) {
+            this.neighbourWeightsMap = new HashMap<>();
+            this.vertexId = vertexId;
+            this.selected = selected;
+            this.neighbours = new BitSet(vertexCount);
         }
-
-        return dot;
+           
+        private void addNeighbour(final int neighbourPosition, final int additionalWeight) {
+            int currentWeight = neighbourWeightsMap.getOrDefault(neighbourPosition, 0);
+            neighbourWeightsMap.put(neighbourPosition,currentWeight + additionalWeight);
+            neighbours.set(neighbourPosition, true);
```

Comment:
for completion, should a recalculation of `magnitude` be done everytime to weights map is updated?